
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
        <meta name="author" content="Kevin Arvai" />
        <meta name="keywords" content="python,genomics,bioinformatics,genotype,genetics" />
        <meta name="description" content="A walkthrough to correctly label genotypes in the gnomAD .vcf genotype count arrays." />

<!-- Open Graph tags -->
<meta property="og:site_name" content="Blogging Bioinformatics"/>
    <meta property="og:type" content="article"/>
    <meta property="og:title" content="Exploring genotype arrays in gnomAD .vcf files"/>
    <meta property="og:url" content="https://arvkevi.github.io/gnomAD_vcf_parsing.html"/>
    <meta property="og:description" content="A walkthrough to correctly label genotypes in the gnomAD .vcf genotype count arrays."/>
    <meta property="article:published_time" content="2017-07-07" />
    <meta property="article:section" content="bioinformatics" />
        <meta property="article:tag" content="python" />
        <meta property="article:tag" content="genomics" />
        <meta property="article:tag" content="bioinformatics" />
        <meta property="article:tag" content="genotype" />
        <meta property="article:tag" content="genetics" />
        <meta property="article:author" content="Kevin Arvai" />

    <title>Exploring genotype arrays in gnomAD .vcf files - Blogging Bioinformatics</title>

        <link rel="stylesheet" href="https://arvkevi.github.io/theme/css/bootstrap.spacelab.min.css" type="text/css"/>

    <link href="https://arvkevi.github.io/theme/css/font-awesome.min.css" rel="stylesheet" />
    <link href="https://arvkevi.github.io/theme/css/pygments/default.css" rel="stylesheet" />
    
    <link href="https://arvkevi.github.io/theme/css/pelican-twitchy.min.css" rel="stylesheet" />

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
        <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
        <script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
    <![endif]-->
    
    <!-- Feeds -->
        <link href="https://arvkevi.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate"
              title="Blogging Bioinformatics ATOM Feed" />
</head>
<body data-spy="scroll" data-target="#scrollspy">
    <div id="wrapper">
        <!-- Sidebar -->
        <div id="sidebar-wrapper-small" class="twitchy-background">
            <ul id="accordion-small" class="sidebar-nav sidebar-nav-small">
                <li>
        <a href="https://arvkevi.github.io" title="Blogging Bioinformatics" class="collapsed">
            <span class="glyphicon glyphicon-home"></span>
        </a>
    </li>
                <li class="nav-divider"></li>
                <li>
        <a href="https://arvkevi.github.io/archives.html" title="Recent Articles" class="collapsed">
            <span class="glyphicon glyphicon-th-list"></span>
        </a>
    </li>
                
                <li class="nav-divider"></li>
                <li>
                    <a data-toggle="collapse" data-parent="#accordion-small" href="#collapse-social-small" title="Social" class="collapsed">
                        <i class="fa fa-users padding-small"></i>
                    </a>
                </li>
                <li class="panel anti-panel"><ul id="collapse-social-small" class="collapse ">
                    <li>
                        <a href="https://github.com/arvkevi" title="github"><i class="fa fa-github-square fa-lg padding-small"></i></a>
                    </li>
                    <li>
                        <a href="https://www.linkedin.com/in/kevinarvai" title="linkedin"><i class="fa fa-linkedin-square fa-lg padding-small"></i></a>
                    </li>
                    <li>
                        <a href="https://twitter.com/arvkevi" title="twitter"><i class="fa fa-twitter-square fa-lg padding-small"></i></a>
                    </li>
                </ul></li>
                <li class="nav-divider"></li>
                <li>
        <a href="#" title="Back to top" class="collapsed">
            <span class="fa fa-arrow-up"></span>
        </a>
    </li>
            </ul>
        </div>
        <div id="sidebar-wrapper" class="twitchy-background">
            <ul id="accordion" class="sidebar-nav">
                <li class="sidebar-brand">
                    <a href="https://arvkevi.github.io/">
                    </a>
                </li>
                    <li>
                        <a href="https://arvkevi.github.io/archives.html">
                            <span class="glyphicon glyphicon-th-list padding-small"></span>
                            Recent Articles
                        </a>
                    </li>
                <li class="nav-divider"></li>
                <li>
                    <a data-toggle="collapse" data-parent="#accordion" href="#collapse-social">
                        <i class="fa fa-users padding-small"></i>
                        Contact
                    </a>
                </li>
                <li class="panel anti-panel"><ul id="collapse-social" class="sidebar_submenu collapse ">
                    <li>
                        <a href="https://github.com/arvkevi" title="github">
                            <i class="fa fa-github-square fa-lg padding-small"></i>
                            github
                        </a>
                    </li>
                    <li>
                        <a href="https://www.linkedin.com/in/kevinarvai" title="linkedin">
                            <i class="fa fa-linkedin-square fa-lg padding-small"></i>
                            linkedin
                        </a>
                    </li>
                    <li>
                        <a href="https://twitter.com/arvkevi" title="twitter">
                            <i class="fa fa-twitter-square fa-lg padding-small"></i>
                            twitter
                        </a>
                    </li>
                </ul></li>
                
                <li class="nav-divider"></li>
                <li>
                    <a data-toggle="collapse" data-parent="#accordion" href="#collapse-pages">
                        <i class="fa fa-folder-open padding-small"></i>
                        Pages
                    </a>
                </li>
                <li class="panel anti-panel"><ul id="collapse-pages" class="sidebar_submenu collapse ">
                </ul></li>
                
                <li class="nav-divider"></li>
                <li>
                    <a data-toggle="collapse" data-parent="#accordion" href="#collapse-categories">
                        <i class="fa fa-folder-open padding-small"></i>
                        Categories
                    </a>
                </li>
                <li class="panel anti-panel"><ul id="collapse-categories" class="sidebar_submenu collapse ">
                    <li class="active">
                        <a href="https://arvkevi.github.io/category/bioinformatics.html">
                            <i class="fa fa-folder-open padding-small"></i>
                            bioinformatics
                            <span class="badge pull-right categorybadge">1</span>
                        </a>
                    </li>
                    <li >
                        <a href="https://arvkevi.github.io/category/computation.html">
                            <i class="fa fa-folder-open padding-small"></i>
                            computation
                            <span class="badge pull-right categorybadge">1</span>
                        </a>
                    </li>
                    <li >
                        <a href="https://arvkevi.github.io/category/lifehack.html">
                            <i class="fa fa-folder-open padding-small"></i>
                            lifehack
                            <span class="badge pull-right categorybadge">1</span>
                        </a>
                    </li>
                </ul></li>
                
            </ul>
        </div>
        <!-- /#sidebar-wrapper -->
        <!-- open/close sidebar -->
        <a href="#menu-toggle" class="btn btn-default" id="menu-toggle">
            <span id="right-arrow" class="glyphicon glyphicon-chevron-right"  title="expand sidebar"></span>
            <span id="left-arrow" class="glyphicon glyphicon-chevron-left" title="minimize sidebar"></span>
        </a>
       <!-- /open/close sidebar -->

        <!-- Page Content -->
        <div id="page-content-wrapper">
            <div class="container-fluid">
<section id="content">
    <article>
        <div class="row">
            <div class="col-lg-9">
                <header class="page-header">
                    <h1>
                        <a href="https://arvkevi.github.io/gnomAD_vcf_parsing.html"
                           rel="bookmark"
                           title="Permalink to Exploring genotype arrays in gnomAD .vcf files">
                            Exploring genotype arrays in gnomAD .vcf files
                        </a>
                        <small>
<div class="post-info">
    <div class="publish-info-block">
        <small>
            <span class="published">
                <i class="fa fa-calendar padding-small"></i><time datetime="2017-07-07T00:00:00-04:00"> Fri 07 July 2017</time>
            </span>
            <span class="category">
                <i class="fa fa-folder-open padding-small"></i><a href="https://arvkevi.github.io/category/bioinformatics.html">bioinformatics</a>
            </span>
            <span class="tags">
                <i class="fa fa-tags padding-small"></i>
                <a href="https://arvkevi.github.io/tag/python.html">python</a> /                 <a href="https://arvkevi.github.io/tag/genomics.html">genomics</a> /                 <a href="https://arvkevi.github.io/tag/bioinformatics.html">bioinformatics</a> /                 <a href="https://arvkevi.github.io/tag/genotype.html">genotype</a> /                 <a href="https://arvkevi.github.io/tag/genetics.html">genetics</a>            </span>
        </small>
    </div>
</div><!-- /.post-info -->                        </small>
                    </h1>
                </header>
            </div>
        </div>
        <div class="row">
            <div class="col-lg-9">
                <div class="entry-content">
                    <html><head><style type="text/css">/*!
*
* IPython notebook
*
*/
/* CSS font colors for translated ANSI colors. */
.ansibold {
  font-weight: bold;
}
/* use dark versions for foreground, to improve visibility */
.ansiblack {
  color: black;
}
.ansired {
  color: darkred;
}
.ansigreen {
  color: darkgreen;
}
.ansiyellow {
  color: #c4a000;
}
.ansiblue {
  color: darkblue;
}
.ansipurple {
  color: darkviolet;
}
.ansicyan {
  color: steelblue;
}
.ansigray {
  color: gray;
}
/* and light for background, for the same reason */
.ansibgblack {
  background-color: black;
}
.ansibgred {
  background-color: red;
}
.ansibggreen {
  background-color: green;
}
.ansibgyellow {
  background-color: yellow;
}
.ansibgblue {
  background-color: blue;
}
.ansibgpurple {
  background-color: magenta;
}
.ansibgcyan {
  background-color: cyan;
}
.ansibggray {
  background-color: gray;
}
div.cell {
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
  border-radius: 2px;
  box-sizing: border-box;
  -moz-box-sizing: border-box;
  -webkit-box-sizing: border-box;
  border-width: 1px;
  border-style: solid;
  border-color: transparent;
  width: 100%;
  padding: 5px;
  /* This acts as a spacer between cells, that is outside the border */
  margin: 0px;
  outline: none;
  border-left-width: 1px;
  padding-left: 5px;
  background: linear-gradient(to right, transparent -40px, transparent 1px, transparent 1px, transparent 100%);
}
div.cell.jupyter-soft-selected {
  border-left-color: #90CAF9;
  border-left-color: #E3F2FD;
  border-left-width: 1px;
  padding-left: 5px;
  border-right-color: #E3F2FD;
  border-right-width: 1px;
  background: #E3F2FD;
}
@media print {
  div.cell.jupyter-soft-selected {
    border-color: transparent;
  }
}
div.cell.selected {
  border-color: #ababab;
  border-left-width: 0px;
  padding-left: 6px;
  background: linear-gradient(to right, #42A5F5 -40px, #42A5F5 5px, transparent 5px, transparent 100%);
}
@media print {
  div.cell.selected {
    border-color: transparent;
  }
}
div.cell.selected.jupyter-soft-selected {
  border-left-width: 0;
  padding-left: 6px;
  background: linear-gradient(to right, #42A5F5 -40px, #42A5F5 7px, #E3F2FD 7px, #E3F2FD 100%);
}
.edit_mode div.cell.selected {
  border-color: #66BB6A;
  border-left-width: 0px;
  padding-left: 6px;
  background: linear-gradient(to right, #66BB6A -40px, #66BB6A 5px, transparent 5px, transparent 100%);
}
@media print {
  .edit_mode div.cell.selected {
    border-color: transparent;
  }
}
.prompt {
  /* This needs to be wide enough for 3 digit prompt numbers: In[100]: */
  min-width: 14ex;
  /* This padding is tuned to match the padding on the CodeMirror editor. */
  padding: 0.4em;
  margin: 0px;
  font-family: monospace;
  text-align: right;
  /* This has to match that of the the CodeMirror class line-height below */
  line-height: 1.21429em;
  /* Don't highlight prompt number selection */
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  /* Use default cursor */
  cursor: default;
}
@media (max-width: 540px) {
  .prompt {
    text-align: left;
  }
}
div.inner_cell {
  min-width: 0;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
  /* Old browsers */
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  box-flex: 1;
  /* Modern browsers */
  flex: 1;
}
/* input_area and input_prompt must match in top border and margin for alignment */
div.input_area {
  border: 1px solid #cfcfcf;
  border-radius: 2px;
  background: #f7f7f7;
  line-height: 1.21429em;
}
/* This is needed so that empty prompt areas can collapse to zero height when there
   is no content in the output_subarea and the prompt. The main purpose of this is
   to make sure that empty JavaScript output_subareas have no height. */
div.prompt:empty {
  padding-top: 0;
  padding-bottom: 0;
}
div.unrecognized_cell {
  padding: 5px 5px 5px 0px;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
div.unrecognized_cell .inner_cell {
  border-radius: 2px;
  padding: 5px;
  font-weight: bold;
  color: red;
  border: 1px solid #cfcfcf;
  background: #eaeaea;
}
div.unrecognized_cell .inner_cell a {
  color: inherit;
  text-decoration: none;
}
div.unrecognized_cell .inner_cell a:hover {
  color: inherit;
  text-decoration: none;
}
@media (max-width: 540px) {
  div.unrecognized_cell > div.prompt {
    display: none;
  }
}
div.code_cell {
  /* avoid page breaking on code cells when printing */
}
@media print {
  div.code_cell {
    page-break-inside: avoid;
  }
}
/* any special styling for code cells that are currently running goes here */
div.input {
  page-break-inside: avoid;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
@media (max-width: 540px) {
  div.input {
    /* Old browsers */
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-box-align: stretch;
    display: -moz-box;
    -moz-box-orient: vertical;
    -moz-box-align: stretch;
    display: box;
    box-orient: vertical;
    box-align: stretch;
    /* Modern browsers */
    display: flex;
    flex-direction: column;
    align-items: stretch;
  }
}
/* input_area and input_prompt must match in top border and margin for alignment */
div.input_prompt {
  color: #303F9F;
  border-top: 1px solid transparent;
}
div.input_area > div.highlight {
  margin: 0.4em;
  border: none;
  padding: 0px;
  background-color: transparent;
}
div.input_area > div.highlight > pre {
  margin: 0px;
  border: none;
  padding: 0px;
  background-color: transparent;
}
/* The following gets added to the <head> if it is detected that the user has a
 * monospace font with inconsistent normal/bold/italic height.  See
 * notebookmain.js.  Such fonts will have keywords vertically offset with
 * respect to the rest of the text.  The user should select a better font.
 * See: https://github.com/ipython/ipython/issues/1503
 *
 * .CodeMirror span {
 *      vertical-align: bottom;
 * }
 */
.CodeMirror {
  line-height: 1.21429em;
  /* Changed from 1em to our global default */
  font-size: 14px;
  height: auto;
  /* Changed to auto to autogrow */
  background: none;
  /* Changed from white to allow our bg to show through */
}
.CodeMirror-scroll {
  /*  The CodeMirror docs are a bit fuzzy on if overflow-y should be hidden or visible.*/
  /*  We have found that if it is visible, vertical scrollbars appear with font size changes.*/
  overflow-y: hidden;
  overflow-x: auto;
}
.CodeMirror-lines {
  /* In CM2, this used to be 0.4em, but in CM3 it went to 4px. We need the em value because */
  /* we have set a different line-height and want this to scale with that. */
  padding: 0.4em;
}
.CodeMirror-linenumber {
  padding: 0 8px 0 4px;
}
.CodeMirror-gutters {
  border-bottom-left-radius: 2px;
  border-top-left-radius: 2px;
}
.CodeMirror pre {
  /* In CM3 this went to 4px from 0 in CM2. We need the 0 value because of how we size */
  /* .CodeMirror-lines */
  padding: 0;
  border: 0;
  border-radius: 0;
}
/*

Original style from softwaremaniacs.org (c) Ivan Sagalaev <Maniac@SoftwareManiacs.Org>
Adapted from GitHub theme

*/
.highlight-base {
  color: #000;
}
.highlight-variable {
  color: #000;
}
.highlight-variable-2 {
  color: #1a1a1a;
}
.highlight-variable-3 {
  color: #333333;
}
.highlight-string {
  color: #BA2121;
}
.highlight-comment {
  color: #408080;
  font-style: italic;
}
.highlight-number {
  color: #080;
}
.highlight-atom {
  color: #88F;
}
.highlight-keyword {
  color: #008000;
  font-weight: bold;
}
.highlight-builtin {
  color: #008000;
}
.highlight-error {
  color: #f00;
}
.highlight-operator {
  color: #AA22FF;
  font-weight: bold;
}
.highlight-meta {
  color: #AA22FF;
}
/* previously not defined, copying from default codemirror */
.highlight-def {
  color: #00f;
}
.highlight-string-2 {
  color: #f50;
}
.highlight-qualifier {
  color: #555;
}
.highlight-bracket {
  color: #997;
}
.highlight-tag {
  color: #170;
}
.highlight-attribute {
  color: #00c;
}
.highlight-header {
  color: blue;
}
.highlight-quote {
  color: #090;
}
.highlight-link {
  color: #00c;
}
/* apply the same style to codemirror */
.cm-s-ipython span.cm-keyword {
  color: #008000;
  font-weight: bold;
}
.cm-s-ipython span.cm-atom {
  color: #88F;
}
.cm-s-ipython span.cm-number {
  color: #080;
}
.cm-s-ipython span.cm-def {
  color: #00f;
}
.cm-s-ipython span.cm-variable {
  color: #000;
}
.cm-s-ipython span.cm-operator {
  color: #AA22FF;
  font-weight: bold;
}
.cm-s-ipython span.cm-variable-2 {
  color: #1a1a1a;
}
.cm-s-ipython span.cm-variable-3 {
  color: #333333;
}
.cm-s-ipython span.cm-comment {
  color: #408080;
  font-style: italic;
}
.cm-s-ipython span.cm-string {
  color: #BA2121;
}
.cm-s-ipython span.cm-string-2 {
  color: #f50;
}
.cm-s-ipython span.cm-meta {
  color: #AA22FF;
}
.cm-s-ipython span.cm-qualifier {
  color: #555;
}
.cm-s-ipython span.cm-builtin {
  color: #008000;
}
.cm-s-ipython span.cm-bracket {
  color: #997;
}
.cm-s-ipython span.cm-tag {
  color: #170;
}
.cm-s-ipython span.cm-attribute {
  color: #00c;
}
.cm-s-ipython span.cm-header {
  color: blue;
}
.cm-s-ipython span.cm-quote {
  color: #090;
}
.cm-s-ipython span.cm-link {
  color: #00c;
}
.cm-s-ipython span.cm-error {
  color: #f00;
}
.cm-s-ipython span.cm-tab {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAMCAYAAAAkuj5RAAAAAXNSR0IArs4c6QAAAGFJREFUSMft1LsRQFAQheHPowAKoACx3IgEKtaEHujDjORSgWTH/ZOdnZOcM/sgk/kFFWY0qV8foQwS4MKBCS3qR6ixBJvElOobYAtivseIE120FaowJPN75GMu8j/LfMwNjh4HUpwg4LUAAAAASUVORK5CYII=);
  background-position: right;
  background-repeat: no-repeat;
}
div.output_wrapper {
  /* this position must be relative to enable descendents to be absolute within it */
  position: relative;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
  z-index: 1;
}
/* class for the output area when it should be height-limited */
div.output_scroll {
  /* ideally, this would be max-height, but FF barfs all over that */
  height: 24em;
  /* FF needs this *and the wrapper* to specify full width, or it will shrinkwrap */
  width: 100%;
  overflow: auto;
  border-radius: 2px;
  -webkit-box-shadow: inset 0 2px 8px rgba(0, 0, 0, 0.8);
  box-shadow: inset 0 2px 8px rgba(0, 0, 0, 0.8);
  display: block;
}
/* output div while it is collapsed */
div.output_collapsed {
  margin: 0px;
  padding: 0px;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
}
div.out_prompt_overlay {
  height: 100%;
  padding: 0px 0.4em;
  position: absolute;
  border-radius: 2px;
}
div.out_prompt_overlay:hover {
  /* use inner shadow to get border that is computed the same on WebKit/FF */
  -webkit-box-shadow: inset 0 0 1px #000;
  box-shadow: inset 0 0 1px #000;
  background: rgba(240, 240, 240, 0.5);
}
div.output_prompt {
  color: #D84315;
}
/* This class is the outer container of all output sections. */
div.output_area {
  padding: 0px;
  page-break-inside: avoid;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
div.output_area .MathJax_Display {
  text-align: left !important;
}
div.output_area 
div.output_area 
div.output_area img,
div.output_area svg {
  max-width: 100%;
  height: auto;
}
div.output_area img.unconfined,
div.output_area svg.unconfined {
  max-width: none;
}
/* This is needed to protect the pre formating from global settings such
   as that of bootstrap */
.output {
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
}
@media (max-width: 540px) {
  div.output_area {
    /* Old browsers */
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-box-align: stretch;
    display: -moz-box;
    -moz-box-orient: vertical;
    -moz-box-align: stretch;
    display: box;
    box-orient: vertical;
    box-align: stretch;
    /* Modern browsers */
    display: flex;
    flex-direction: column;
    align-items: stretch;
  }
}
div.output_area pre {
  margin: 0;
  padding: 0;
  border: 0;
  vertical-align: baseline;
  color: black;
  background-color: transparent;
  border-radius: 0;
}
/* This class is for the output subarea inside the output_area and after
   the prompt div. */
div.output_subarea {
  overflow-x: auto;
  padding: 0.4em;
  /* Old browsers */
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  box-flex: 1;
  /* Modern browsers */
  flex: 1;
  max-width: calc(100% - 14ex);
}
div.output_scroll div.output_subarea {
  overflow-x: visible;
}
/* The rest of the output_* classes are for special styling of the different
   output types */
/* all text output has this class: */
div.output_text {
  text-align: left;
  color: #000;
  /* This has to match that of the the CodeMirror class line-height below */
  line-height: 1.21429em;
}
/* stdout/stderr are 'text' as well as 'stream', but execute_result/error are *not* streams */
div.output_stderr {
  background: #fdd;
  /* very light red background for stderr */
}
div.output_latex {
  text-align: left;
}
/* Empty output_javascript divs should have no height */
div.output_javascript:empty {
  padding: 0;
}
.js-error {
  color: darkred;
}
/* raw_input styles */
div.raw_input_container {
  line-height: 1.21429em;
  padding-top: 5px;
}
pre.raw_input_prompt {
  /* nothing needed here. */
}
input.raw_input {
  font-family: monospace;
  font-size: inherit;
  color: inherit;
  width: auto;
  /* make sure input baseline aligns with prompt */
  vertical-align: baseline;
  /* padding + margin = 0.5em between prompt and cursor */
  padding: 0em 0.25em;
  margin: 0em 0.25em;
}
input.raw_input:focus {
  box-shadow: none;
}
p.p-space {
  margin-bottom: 10px;
}
div.output_unrecognized {
  padding: 5px;
  font-weight: bold;
  color: red;
}
div.output_unrecognized a {
  color: inherit;
  text-decoration: none;
}
div.output_unrecognized a:hover {
  color: inherit;
  text-decoration: none;
}
.rendered_html {
  color: #000;
  /* any extras will just be numbers: */
}



.rendered_html :link {
  text-decoration: underline;
}
.rendered_html :visited {
  text-decoration: underline;
}






.rendered_html h1:first-child {
  margin-top: 0.538em;
}
.rendered_html h2:first-child {
  margin-top: 0.636em;
}
.rendered_html h3:first-child {
  margin-top: 0.777em;
}
.rendered_html h4:first-child {
  margin-top: 1em;
}
.rendered_html h5:first-child {
  margin-top: 1em;
}
.rendered_html h6:first-child {
  margin-top: 1em;
}








.rendered_html * + ul {
  margin-top: 1em;
}
.rendered_html * + ol {
  margin-top: 1em;
}


.rendered_html pre,



.rendered_html tr,
.rendered_html th,

.rendered_html td,


.rendered_html * + table {
  margin-top: 1em;
}

.rendered_html * + p {
  margin-top: 1em;
}

.rendered_html * + img {
  margin-top: 1em;
}
.rendered_html img,

.rendered_html img.unconfined,

div.text_cell {
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
@media (max-width: 540px) {
  div.text_cell > div.prompt {
    display: none;
  }
}
div.text_cell_render {
  /*font-family: "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;*/
  outline: none;
  resize: none;
  width: inherit;
  border-style: none;
  padding: 0.5em 0.5em 0.5em 0.4em;
  color: #000;
  box-sizing: border-box;
  -moz-box-sizing: border-box;
  -webkit-box-sizing: border-box;
}
a.anchor-link:link {
  text-decoration: none;
  padding: 0px 20px;
  visibility: hidden;
}
h1:hover .anchor-link,
h2:hover .anchor-link,
h3:hover .anchor-link,
h4:hover .anchor-link,
h5:hover .anchor-link,
h6:hover .anchor-link {
  visibility: visible;
}
.text_cell.rendered .input_area {
  display: none;
}
.text_cell.rendered 
.text_cell.unrendered .text_cell_render {
  display: none;
}
.cm-header-1,
.cm-header-2,
.cm-header-3,
.cm-header-4,
.cm-header-5,
.cm-header-6 {
  font-weight: bold;
  font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
}
.cm-header-1 {
  font-size: 185.7%;
}
.cm-header-2 {
  font-size: 157.1%;
}
.cm-header-3 {
  font-size: 128.6%;
}
.cm-header-4 {
  font-size: 110%;
}
.cm-header-5 {
  font-size: 100%;
  font-style: italic;
}
.cm-header-6 {
  font-size: 100%;
  font-style: italic;
}
</style>
<style type="text/css">.highlight .hll { background-color: #ffffcc }
.highlight  { background: #f8f8f8; }
.highlight .c { color: #408080; font-style: italic } /* Comment */
.highlight .err { border: 1px solid #FF0000 } /* Error */
.highlight .k { color: #008000; font-weight: bold } /* Keyword */
.highlight .o { color: #666666 } /* Operator */
.highlight .ch { color: #408080; font-style: italic } /* Comment.Hashbang */
.highlight .cm { color: #408080; font-style: italic } /* Comment.Multiline */
.highlight .cp { color: #BC7A00 } /* Comment.Preproc */
.highlight .cpf { color: #408080; font-style: italic } /* Comment.PreprocFile */
.highlight .c1 { color: #408080; font-style: italic } /* Comment.Single */
.highlight .cs { color: #408080; font-style: italic } /* Comment.Special */
.highlight .gd { color: #A00000 } /* Generic.Deleted */
.highlight .ge { font-style: italic } /* Generic.Emph */
.highlight .gr { color: #FF0000 } /* Generic.Error */
.highlight .gh { color: #000080; font-weight: bold } /* Generic.Heading */
.highlight .gi { color: #00A000 } /* Generic.Inserted */
.highlight .go { color: #888888 } /* Generic.Output */
.highlight .gp { color: #000080; font-weight: bold } /* Generic.Prompt */
.highlight .gs { font-weight: bold } /* Generic.Strong */
.highlight .gu { color: #800080; font-weight: bold } /* Generic.Subheading */
.highlight .gt { color: #0044DD } /* Generic.Traceback */
.highlight .kc { color: #008000; font-weight: bold } /* Keyword.Constant */
.highlight .kd { color: #008000; font-weight: bold } /* Keyword.Declaration */
.highlight .kn { color: #008000; font-weight: bold } /* Keyword.Namespace */
.highlight .kp { color: #008000 } /* Keyword.Pseudo */
.highlight .kr { color: #008000; font-weight: bold } /* Keyword.Reserved */
.highlight .kt { color: #B00040 } /* Keyword.Type */
.highlight .m { color: #666666 } /* Literal.Number */
.highlight .s { color: #BA2121 } /* Literal.String */
.highlight .na { color: #7D9029 } /* Name.Attribute */
.highlight .nb { color: #008000 } /* Name.Builtin */
.highlight .nc { color: #0000FF; font-weight: bold } /* Name.Class */
.highlight .no { color: #880000 } /* Name.Constant */
.highlight .nd { color: #AA22FF } /* Name.Decorator */
.highlight .ni { color: #999999; font-weight: bold } /* Name.Entity */
.highlight .ne { color: #D2413A; font-weight: bold } /* Name.Exception */
.highlight .nf { color: #0000FF } /* Name.Function */
.highlight .nl { color: #A0A000 } /* Name.Label */
.highlight .nn { color: #0000FF; font-weight: bold } /* Name.Namespace */
.highlight .nt { color: #008000; font-weight: bold } /* Name.Tag */
.highlight .nv { color: #19177C } /* Name.Variable */
.highlight .ow { color: #AA22FF; font-weight: bold } /* Operator.Word */
.highlight .w { color: #bbbbbb } /* Text.Whitespace */
.highlight .mb { color: #666666 } /* Literal.Number.Bin */
.highlight .mf { color: #666666 } /* Literal.Number.Float */
.highlight .mh { color: #666666 } /* Literal.Number.Hex */
.highlight .mi { color: #666666 } /* Literal.Number.Integer */
.highlight .mo { color: #666666 } /* Literal.Number.Oct */
.highlight .sa { color: #BA2121 } /* Literal.String.Affix */
.highlight .sb { color: #BA2121 } /* Literal.String.Backtick */
.highlight .sc { color: #BA2121 } /* Literal.String.Char */
.highlight .dl { color: #BA2121 } /* Literal.String.Delimiter */
.highlight .sd { color: #BA2121; font-style: italic } /* Literal.String.Doc */
.highlight .s2 { color: #BA2121 } /* Literal.String.Double */
.highlight .se { color: #BB6622; font-weight: bold } /* Literal.String.Escape */
.highlight .sh { color: #BA2121 } /* Literal.String.Heredoc */
.highlight .si { color: #BB6688; font-weight: bold } /* Literal.String.Interpol */
.highlight .sx { color: #008000 } /* Literal.String.Other */
.highlight .sr { color: #BB6688 } /* Literal.String.Regex */
.highlight .s1 { color: #BA2121 } /* Literal.String.Single */
.highlight .ss { color: #19177C } /* Literal.String.Symbol */
.highlight .bp { color: #008000 } /* Name.Builtin.Pseudo */
.highlight .fm { color: #0000FF } /* Name.Function.Magic */
.highlight .vc { color: #19177C } /* Name.Variable.Class */
.highlight .vg { color: #19177C } /* Name.Variable.Global */
.highlight .vi { color: #19177C } /* Name.Variable.Instance */
.highlight .vm { color: #19177C } /* Name.Variable.Magic */
.highlight .il { color: #666666 } /* Literal.Number.Integer.Long */</style>
<style type="text/css">
/* Temporary definitions which will become obsolete with Notebook release 5.0 */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-bold { font-weight: bold; }
</style>
</head><body><div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Parsing-genotype-arrays-in-gnomAD-.vcf-files-without-hail">Parsing genotype arrays in gnomAD .vcf files without hail<a class="anchor-link" href="#Parsing-genotype-arrays-in-gnomAD-.vcf-files-without-hail">¶</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>There have been a few <a href="https://macarthurlab.org/2017/02/27/the-genome-aggregation-database-gnomad/">documented changes to gnomAD vcf</a>.  This excerpt describes one of the biggest changes, the genotype count arrays:</p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<blockquote><p>"We have removed the Het fields as they were hard to parse and instead provide genotype counts (GC), or a G-indexed array of counts of individuals per genotype. As mentioned in the Variant QC section, we have fundamentally switched from a site-level filtering strategy to an allele-specific filtering strategy."</p>
</blockquote>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>While the <em>order</em> of the genotypes in these "G-indexed" arrays can be easily inferred for biallelic sites, it becomes more difficult to deduce at multiallelic sites.  My co-worker, Julie, deserves credit for the inspiration for this post and special thanks for the back-and-forth conversation before she arrived at a working solution. We think using <a href="https://github.com/hail-is/hail">hail</a> would alleviate the problem because it was the software used to work with the gnomAD data and it has an extensive framework for working with variants. <strong>The solution outlined below is a pure Python method to assign genotype labels to counts from a raw gnomAD .vcf</strong>.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To follow along, grab the chromosome 22 vcf and index file (or from the <a href="http://gnomad.broadinstitute.org/downloads">gnomAD downloads page</a>)</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [ ]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># grab the gnomAD chromosome 22 vcf.</span>
<span class="o">!</span> wget https://storage.googleapis.com/gnomad-public/release-170228/coverage/genomes/gnomad.chr22.cov.txt.gz
<span class="o">!</span> wget https://storage.googleapis.com/gnomad-public/release-170228/coverage/genomes/gnomad.chr22.cov.txt.gz.tbi
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>I'll be using Python to decode the genotype counts, in particular leaning on <a href="https://github.com/brentp/cyvcf2">cyvcf2</a> for variant parsing and <a href="https://numpy.org">numpy</a> for working with the genotype count data.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [1]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">cyvcf2</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span> <span class="c1"># for visualization in the notebook.</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [2]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># cyvcf2 wraps htslib for efficient parsing:</span>
<span class="n">gnomad22vcf</span> <span class="o">=</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="s1">'gnomad.genomes.r2.0.1.sites.22.vcf.gz'</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Here's a generator function that will return a variant from the gnomAD .vcf based on some filtering criteria.  <code>variant_generator</code> simply iterates through the gnomAD .vcf one record per iteration.  The <code>multiallelic</code> argument serves as a boolean flag because we'll be working with both biallelic and multiallelic variants below.  The <code>alt_af</code> parameter ensures that the returned variant will have some allele diversity. Rare variants serve as poor examples because some alternate alleles will have <code>0</code>, so there won't be any alleles to count.  The <code>variant_generator</code> function isn't intended to be sophisticated or useful beyond the scope of this blog post, it just pulls relevant variants for us to explore.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [3]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">variant_generator</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span> <span class="n">multiallelic</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">alt_af</span><span class="o">=</span><span class="mf">0.05</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Iterate through .vcf file.</span>
<span class="sd">    Skip FILTER variants</span>
<span class="sd">    restrict to missense.</span>
<span class="sd">    Return variants the exceed some alternate allele frequency <alt_af>.</alt_af></span>
<span class="sd">    """</span>
    <span class="k">for</span> <span class="n">_var</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">_var</span><span class="o">.</span><span class="n">FILTER</span><span class="p">:</span>
            <span class="c1"># skip FILTER variants</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="s1">'missense'</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">_var</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'CSQ'</span><span class="p">]:</span>
            <span class="c1"># do your best to skip non-exonic variants</span>
            <span class="k">continue</span>
            
        <span class="k">if</span> <span class="n">multiallelic</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">_var</span><span class="o">.</span><span class="n">ALT</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span>
                <span class="nb">any</span><span class="p">(</span><span class="mf">0.4</span> <span class="o">&gt;</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="n">alt_af</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_var</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'AF'</span><span class="p">])):</span>
                <span class="k">yield</span> <span class="n">_var</span>
        <span class="k">else</span><span class="p">:</span> <span class="c1"># biallelic</span>
            <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">_var</span><span class="o">.</span><span class="n">ALT</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span>
                <span class="n">_var</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'AF'</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">alt_af</span> <span class="o">&lt;</span> <span class="mf">0.4</span><span class="p">):</span>
                <span class="k">yield</span> <span class="n">_var</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>For posterity, we'll only look at <strong>one</strong> of the global sub-popuplations. Non-Finnish European, denoted by <code>NFE</code>, is a good choice for this exercise because that group has a large sample size.  The first example will start with a biallelic variant, as these have less complexity than multiallelic counterparts.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [4]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">variant</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">variant_generator</span><span class="p">(</span><span class="n">gnomad22vcf</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [5]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">variant</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[5]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>Variant(22:16449050 G/A)</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="gnomAD-bialleleic-genotypes-breakdown">gnomAD bialleleic genotypes breakdown<a class="anchor-link" href="#gnomAD-bialleleic-genotypes-breakdown">¶</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In the gnomAD .vcf, the <code>GC_NFE</code> field shows the aforementioned "G-indexed" array (from the gnomAD blog post). The order of "each genotype" is not immediately obvious, but we can infer the order by visiting the <a href="http://gnomad.broadinstitute.org/variant/22-16449050-G-A">variant page</a> (and/or by calculating below).</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [7]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[7]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>(3180, 3430, 720)</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>These are genotype counts. In short, the genotype at a given loci is defined by the allele composition.  For this particular variant the possible <strong>alleles</strong> are <strong><code>G</code></strong> and <strong><code>A</code></strong> and the possbible <strong>genotypes</strong> for those alleles are the combinations of the alleles. The Punnett square figure below provides a visual representation of the difference between a genotype and allele, think of <strong><code>A</code></strong> and <strong><code>G</code></strong> instead of <strong><code>B</code></strong> and <strong><code>b</code></strong>.<br/>
<img alt="genotypes" class="img-responsive" src="http://pad3.whstatic.com/images/thumb/0/0e/Use-a-Punnett-Square-to-Do-a-Monohybrid-Cross-Step-5-Version-2.jpg/aid38087-v4-728px-Use-a-Punnett-Square-to-Do-a-Monohybrid-Cross-Step-5-Version-2.jpg.webp"/>
<strong>image from: <a href="http://www.wikihow.com/Use-a-Punnett-Square-to-Do-a-Monohybrid-Cross">http://www.wikihow.com/Use-a-Punnett-Square-to-Do-a-Monohybrid-Cross</a></strong></p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>One observation is that there are only 3 counts in the genotype count array, but there are four cells in the Punnett square.  One of those counts represents both heterozygous genotypes, those two orange cells in the square. The question we're interested in is <strong>can we reproduce the alternate allele frequency from the <em>genotype counts array</em>?</strong></p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [8]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># alternate allele count for NFE</span>
<span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'AC_NFE'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[8]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>4870</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [9]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># same genotype counts array from above</span>
<span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[9]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>(3180, 3430, 720)</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Take the second second element and add it with double the third element. The third element is doubled because
we're assuming it represents homozygous alternate allele state, or <strong><code>AA</code></strong>.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [10]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="mi">3430</span> <span class="o">+</span> <span class="p">(</span><span class="mi">720</span><span class="o">*</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[10]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>4870</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>So far it appears that the second element represents heterozygous variant counts and the third element represents homozygous variant counts. Let's check that the <code>AN_NFE</code> matches twice the nubmer of all the genotypes in <code>GC_NFE</code>.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [11]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># sum up the genotypes array from the variant record &amp; multiply by 2</span>
<span class="n">alleleN</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">])</span> <span class="o">*</span> <span class="mi">2</span>
<span class="nb">print</span><span class="p">(</span><span class="n">alleleN</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>14660
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Does the manual sum equal the gnomad .vcf <code>AN_NFE</code> (Allele Number)?</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [12]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'AN_NFE'</span><span class="p">]</span> <span class="o">==</span> <span class="n">alleleN</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[12]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>True</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now manually check the allele frequency.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [13]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="p">(</span><span class="mi">3430</span> <span class="o">+</span> <span class="p">(</span><span class="mi">720</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span> <span class="o">/</span> <span class="n">alleleN</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[13]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>0.3321964529331514</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [14]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'AF_NFE'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[14]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>0.3321959972381592</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Great, push ahead under the assumption that the genotype counts array is correct ~&gt; <strong>(hom_ref, het_alt, hom_alt)</strong></p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [15]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">ref</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">REF</span>
<span class="n">alt</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">ALT</span>
<span class="n">gcounts_array</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gcounts_array</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>(3180, 3430, 720)
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [16]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">alleles</span> <span class="o">=</span> <span class="n">alt</span>
<span class="n">alleles</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ref</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>['G', 'A']
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Set up a pandas DataFrame to visualize the genotypes and their counts  -- this will resemble the Punnett square above.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [17]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))),</span> <span class="n">index</span><span class="o">=</span><span class="n">alleles</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">alleles</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [18]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">n_alleles</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span>
<span class="n">allele_order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tril_indices</span><span class="p">(</span><span class="n">n_alleles</span><span class="p">)</span>
<span class="c1">#https://docs.scipy.org/doc/numpy/reference/generated/numpy.tril_indices.html</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [19]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">allele_order</span><span class="p">]</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">]</span>
<span class="n">df</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[19]:</div>
<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<table border="1" class="table-hover  table-striped dataframe table">
<thead>
<tr style="text-align: right;">
<th></th>
<th>G</th>
<th>A</th>
</tr>
</thead>
<tbody>
<tr>
<th>G</th>
<td>3180.0</td>
<td>0.0</td>
</tr>
<tr>
<th>A</th>
<td>3430.0</td>
<td>720.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="gnomAD-multialleleic-genotypes-breakdown">gnomAD multialleleic genotypes breakdown<a class="anchor-link" href="#gnomAD-multialleleic-genotypes-breakdown">¶</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The biallelic variant was fairly straight-forward. Where this array conversion really comes in handy is the multiallelic sites.  Use the <code>variant_generator</code> to grab the next variant, but change <strong><code>multiallelic=True</code></strong>.  Also, loop through the .vcf until we reach a variant with more than 2 alternate alleles, this will help show the benefit of deconstructing the added complexity that comes with multiallelic sites.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [20]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># will run for a minute or so</span>
<span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variant_generator</span><span class="p">(</span><span class="n">gnomad22vcf</span><span class="p">,</span> <span class="n">multiallelic</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">ALT</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">break</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [21]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">variant</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[21]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>Variant(22:46449891 G/A,C,T)</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [22]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">ref</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">REF</span>
<span class="n">alt</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">ALT</span>
<span class="n">gcounts_array</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gcounts_array</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>(2871, 3197, 935, 252, 136, 2, 0, 2, 0, 0)
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>That got confusing fast!  The numpy method, <code>np.tril_indices</code>, trick should help us obtain accurate genotypes.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [23]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># same code as above -- just applied to a new (multiallelic) variant</span>
<span class="n">alleles</span> <span class="o">=</span> <span class="n">alt</span>
<span class="n">alleles</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ref</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))),</span> <span class="n">index</span><span class="o">=</span><span class="n">alleles</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">alleles</span><span class="p">)</span>

<span class="n">n_alleles</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">)</span>
<span class="n">allele_order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tril_indices</span><span class="p">(</span><span class="n">n_alleles</span><span class="p">)</span>

<span class="n">df</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">allele_order</span><span class="p">]</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>['G', 'A', 'C', 'T']
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>There are <strong><code>5,205</code></strong> <strong><code>A</code></strong> alleles in <code>NFE</code> at this site according to the <a href="http://gnomad.broadinstitute.org/variant/22-46449891-G-A">gnomAD variant page for G&gt;A</a>. Add up the cells with <strong><code>A</code></strong> alleles, remembering to double the homozgyous cell.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [24]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[(</span><span class="s1">'background: lightgreen'</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s1">'A'</span> <span class="ow">and</span> <span class="n">_</span> <span class="o">&gt;</span> <span class="mi">0</span>
                            <span class="k">else</span> <span class="p">(</span><span class="s1">'background: lightgreen'</span> <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="s1">'A'</span> <span class="ow">and</span> <span class="n">_</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">''</span><span class="p">))</span>
                            <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">_</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()])</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[24]:</div>
<div class="output_html rendered_html output_subarea output_execute_result">
<style type="text/css">
        
        
            #T_6de0532c_60c5_11e7_8ecb_42010a8e0002row1_col0 {
            
                background:  lightgreen;
            
            }
        
            #T_6de0532c_60c5_11e7_8ecb_42010a8e0002row1_col1 {
            
                background:  lightgreen;
            
            }
        
            #T_6de0532c_60c5_11e7_8ecb_42010a8e0002row2_col1 {
            
                background:  lightgreen;
            
            }
        
            #T_6de0532c_60c5_11e7_8ecb_42010a8e0002row3_col1 {
            
                background:  lightgreen;
            
            }
        
        </style>
<table class="table-hover  table-striped table" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002" none="">
<thead>
<tr>
<th class="blank level0">
</th><th class="col_heading level0 col0" colspan="1">
                  G
                
                
                
                </th><th class="col_heading level0 col1" colspan="1">
                  A
                
                
                
                </th><th class="col_heading level0 col2" colspan="1">
                  C
                
                
                
                </th><th class="col_heading level0 col3" colspan="1">
                  T
                
                
            </th></tr>
</thead>
<tbody>
<tr>
<th class="row_heading level0 row0" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002" rowspan="1">
                    G
                
                
                
                </th><td class="data row0 col0" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row0_col0">
                    2871
                
                
                
                </td><td class="data row0 col1" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row0_col1">
                    0
                
                
                
                </td><td class="data row0 col2" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row0_col2">
                    0
                
                
                
                </td><td class="data row0 col3" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row0_col3">
                    0
                
                
            </td></tr>
<tr>
<th class="row_heading level0 row1" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002" rowspan="1">
                    A
                
                
                
                </th><td class="data row1 col0" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row1_col0">
                    3197
                
                
                
                </td><td class="data row1 col1" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row1_col1">
                    935
                
                
                
                </td><td class="data row1 col2" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row1_col2">
                    0
                
                
                
                </td><td class="data row1 col3" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row1_col3">
                    0
                
                
            </td></tr>
<tr>
<th class="row_heading level0 row2" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002" rowspan="1">
                    C
                
                
                
                </th><td class="data row2 col0" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row2_col0">
                    252
                
                
                
                </td><td class="data row2 col1" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row2_col1">
                    136
                
                
                
                </td><td class="data row2 col2" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row2_col2">
                    2
                
                
                
                </td><td class="data row2 col3" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row2_col3">
                    0
                
                
            </td></tr>
<tr>
<th class="row_heading level0 row3" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002" rowspan="1">
                    T
                
                
                
                </th><td class="data row3 col0" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row3_col0">
                    0
                
                
                
                </td><td class="data row3 col1" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row3_col1">
                    2
                
                
                
                </td><td class="data row3 col2" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row3_col2">
                    0
                
                
                
                </td><td class="data row3 col3" id="T_6de0532c_60c5_11e7_8ecb_42010a8e0002row3_col3">
                    0
                
                
            </td></tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [25]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># manual calculation</span>
<span class="mi">3197</span> <span class="o">+</span> <span class="p">(</span><span class="mi">935</span><span class="p">)</span><span class="o">*</span><span class="mi">2</span> <span class="o">+</span> <span class="mi">136</span> <span class="o">+</span> <span class="mi">2</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[25]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>5205</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>That's accurate, to check other's you can use the three gnomAD variant pages to check that the allele counts match.<br/>
<a href="http://gnomad.broadinstitute.org/variant/22-46449891-G-A">G&gt;A</a><br/>
<a href="http://gnomad.broadinstitute.org/variant/22-46449891-G-C">G&gt;C</a><br/>
<a href="http://gnomad.broadinstitute.org/variant/22-46449891-G-T">G&gt;T</a></p>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This post can really be distilled down to assigning genotype labels to the genotype count array from the gnomAD vcf.  Here's a bare-bones function to do that.</p>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [26]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">gnomAD_Gindex</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">):</span>
    <span class="sd">"""</span>
<span class="sd">    Assigns genotype labels to a gnomAD genotype count arrray.</span>
<span class="sd">    """</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="n">alleles</span> <span class="o">=</span> <span class="n">alt</span>
    <span class="n">alleles</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ref</span><span class="p">)</span>
    
    <span class="n">genotype_labels</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">a1</span><span class="p">,</span><span class="n">a2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">tril_indices</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alleles</span><span class="p">))):</span>
        <span class="n">genotype_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alleles</span><span class="p">[</span><span class="n">a1</span><span class="p">]</span> <span class="o">+</span> <span class="n">alleles</span><span class="p">[</span><span class="n">a2</span><span class="p">])</span>
        
    <span class="k">return</span> <span class="n">genotype_labels</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [27]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gt_labels</span> <span class="o">=</span> <span class="n">gnomAD_Gindex</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">REF</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">ALT</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [28]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gt_labels</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt output_prompt">Out[28]:</div>
<div class="output_text output_subarea output_execute_result">
<pre>['GG', 'AG', 'AA', 'CG', 'CA', 'CC', 'TG', 'TA', 'TC', 'TT']</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In [29]:</div>
<div class="inner_cell">
<div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">gt</span><span class="p">,</span> <span class="n">gtcount</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">gt_labels</span><span class="p">,</span> <span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="p">[</span><span class="s1">'GC_NFE'</span><span class="p">]):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">'genotype </span><span class="si">{}</span><span class="s1"> has </span><span class="si">{}</span><span class="s1"> individuals'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gt</span><span class="p">,</span> <span class="n">gtcount</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="prompt"></div>
<div class="output_subarea output_stream output_stdout output_text">
<pre>genotype GG has 2871 individuals
genotype AG has 3197 individuals
genotype AA has 935 individuals
genotype CG has 252 individuals
genotype CA has 136 individuals
genotype CC has 2 individuals
genotype TG has 0 individuals
genotype TA has 2 individuals
genotype TC has 0 individuals
genotype TT has 0 individuals
</pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Comments are welcome.  This was an exploratory exercise over a long weekend, so there could be elements I'm overlooking.  Thanks for reading!</p>
</div>
</div>
</div>

<script type="text/javascript">if (!document.getElementById('mathjaxscript_pelican_#%@#$@#')) {
    var mathjaxscript = document.createElement('script');
    mathjaxscript.id = 'mathjaxscript_pelican_#%@#$@#';
    mathjaxscript.type = 'text/javascript';
    mathjaxscript.src = '//cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML';
    mathjaxscript[(window.opera ? "innerHTML" : "text")] =
        "MathJax.Hub.Config({" +
        "    config: ['MMLorHTML.js']," +
        "    TeX: { extensions: ['AMSmath.js','AMSsymbols.js','noErrors.js','noUndefined.js'], equationNumbers: { autoNumber: 'AMS' } }," +
        "    jax: ['input/TeX','input/MathML','output/HTML-CSS']," +
        "    extensions: ['tex2jax.js','mml2jax.js','MathMenu.js','MathZoom.js']," +
        "    displayAlign: 'center'," +
        "    displayIndent: '0em'," +
        "    showMathMenu: true," +
        "    tex2jax: { " +
        "        inlineMath: [ ['$','$'] ], " +
        "        displayMath: [ ['$$','$$'] ]," +
        "        processEscapes: true," +
        "        preview: 'TeX'," +
        "    }, " +
        "    'HTML-CSS': { " +
        " linebreaks: { automatic: true, width: '95% container' }, " +
        "        styles: { '.MathJax_Display, .MathJax .mo, .MathJax .mi, .MathJax .mn': {color: 'black ! important'} }" +
        "    } " +
        "}); ";
    (document.body || document.getElementsByTagName('head')[0]).appendChild(mathjaxscript);
}
</script>
</body></html>
                </div>
                <footer class="text-right">
                    <p>- Kevin Arvai</p>
                </footer>
    <section id="comments" class="comments ">
        <hr/>
        <h2>Comments</h2>
        <div id="disqus_thread"></div>
        <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by
            Disqus.</a></noscript>
        <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
    </section>
            </div>
        </div>
    </article>
</section>
<footer>
    <hr>
    <div class="row">
        <div class="col-lg-9 text-center">
            <p><small>
                Built by <a href="http://docs.getpelican.com/en/latest">Pelican</a> / <a href="https://github.com/ingwinlu/pelican-twitchy">pelican-twitchy</a>
                &middot;                    &copy; 2017 Kevin Arvai
            </small></p>
        </div>
    </div>
</footer>            </div>
        </div>
        <!-- /#page-content-wrapper -->
    </div>
    <!-- /#wrapper -->
    <!-- jQuery Version 1.11.2 -->
    <script src="https://arvkevi.github.io/theme/js/jquery-1.11.2.min.js"></script>
    <!-- Bootstrap Core JavaScript -->
    <script src="https://arvkevi.github.io/theme/js/bootstrap.min.js"></script>
    <!-- twitchy Script -->
    <script src="https://arvkevi.github.io/theme/js/pelican_twitchy.min.js"></script>
    <!-- Google Analytics -->
    <script type="text/javascript">

        var _gaq = _gaq || [];
        _gaq.push(['_setAccount', 'UA-66112939-4']);
        _gaq.push(['_trackPageview']);

        (function () {
            var ga = document.createElement('script');
            ga.type = 'text/javascript';
            ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0];
            s.parentNode.insertBefore(ga, s);
        })();
    </script>
    <!-- End Google Analytics Code -->

<!-- disqus -->
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'arvkevi-github-io'; // required: replace example with your forum shortname

            var disqus_identifier = 'gnomAD_vcf_parsing';
        var disqus_url = 'https://arvkevi.github.io/gnomAD_vcf_parsing.html';

    var disqus_config = function () {
        this.language = "en";
    };
        /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var dsq = document.createElement('script');
        dsq.type = 'text/javascript';
        dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<!-- /disqus -->
</body>
</html>